hyperspy/axes.py

```diff
@@ -559,22 +559,50 @@ def value2index(self, value, rounding=round):
         ------
         ValueError
             If any value is out of the axis limits.
-
+            If value is NaN
+            If value is an array containing at least 1 NaN value
+            If value is an array containing at least 1 out-of-bound value
         """
         if value is None:
             return None
-        if self.low_value <= value <= self.high_value:
-            return (np.abs(self.axis - value)).argmin()
-        else:
-            index = int(value)
-            if self.size > index >= 0:
-                return index
+
+        #Should evaluate on both arrays and scalars. Raises error if there are
+        #nan values in array
+        if np.all((value >= self.low_value)*(value <= self.high_value)):
+            #Only if all values will evaluate correctly do we implement rounding
+            #function. Rounding functions will strictly operate on numpy arrays
+            #and only evaluate self.axis - v input, v a scalar within value.
+            if rounding is round:
+                #Use argmin(abs) which will return the closest value
+                rounding_index = lambda x: np.abs(x).argmin()
+            elif rounding is math.ceil:
+                #Ceiling means finding index of the closest xi with xi - v >= 0
+                #we look for argmin of strictly non-negative part of self.axis-v.
+                #The trick is to replace strictly negative values with +np.inf
+                rounding_index = lambda x: np.where(x<0,+np.inf,x).argmin()
+            elif rounding is math.floor:
+                #flooring means finding index of the closest xi with xi - v <= 0
+                #we look for armgax of strictly non-positive part of self.axis-v.
+                #The trick is to replace strictly positive values with -np.inf
+                rounding_index = lambda x: np.where(x>0,-np.inf,x).argmax()
             else:
                 raise ValueError(
-                    f'The value {value} is out of the limits '
-                    f'[{self.low_value:.3g}-{self.high_value:.3g}] of the '
-                    f'"{self._get_name()}" axis.'
+                    f'{rounding} is not a supported rounding function. Use '
+                    f'round, math.ceil or math.floor'
                     )
+            #initialise the index same dimension as input, force type to int
+            index = np.empty_like(value,dtype=int)
+            #assign on flat, iterate on flat.
+            for i,v in enumerate(np.asarray(value).flat):
+                index.flat[i] = rounding_index(self.axis - v)
+            #Squeezing to get a scalar out if scalar in. See squeeze doc
+            return np.squeeze(index)[()]
+        else:
+            raise ValueError(
+                f'The value {value} is out of the limits '
```

Comment:
Can you add a test for this line (see codecov warning)?

Reply:
Looks weird to me. This is the case where a value is out of axis and is tested here:

https://github.com/LMSC-NTappy/hyperspy/blob/2e757e940dd11b0d913cb1c78348beb54c133f63/hyperspy/tests/axes/test_data_axis.py#L332-L336

https://github.com/LMSC-NTappy/hyperspy/blob/2e757e940dd11b0d913cb1c78348beb54c133f63/hyperspy/tests/axes/test_data_axis.py#L348-L355